### PR TITLE
Closes pkp/pkp-lib#6539 Extended reCaptcha to the login page.

### PR DESCRIPTION
**Pull request by jonasraoni**
Hi!

Just refactored a bit the validator and added extra error messages.

I've also removed the `classes/webservice` folder. As I've ended up there and didn't find references to it anywhere (the code also has a broken reference).

### CODE REVIEW

File: templates/frontend/pages/userLogin.tpl [hunk 1]
```diff
@@ -73,6 +73,18 @@
 					</span>
 				</label>
 			</div>
+
+			{* recaptcha spam blocker *}
+			{if $recaptchaPublicKey}
+				<fieldset class="recaptcha_wrapper">
+					<div class="fields">
+						<div class="recaptcha">
+							<div class="g-recaptcha" data-sitekey="{$recaptchaPublicKey}"></div>
```
Comment (asmecher) [inline templates/frontend/pages/userLogin.tpl:82]: I suggest applying the `escape` Smarty modifier to `$recaptchaPublicKey`.

File: classes/form/validation/FormValidatorReCaptcha.inc.php [hunk 2]
```diff
@@ -56,59 +56,72 @@ function isValid() {
...
 		}
 
-	}
-}
+		$privateKey = Config::getVar('captcha', 'recaptcha_private_key');
+		if (empty($privateKey)) {
+			throw new Exception('common.captcha.error.invalid-input-secret');
+		}
 
+		$httpClient = Application::get()->getHttpClient();
```
Comment (asmecher) [inline classes/form/validation/FormValidatorReCaptcha.inc.php:92]: Happy to see the expanded use of the Guzzle client!

File: classes/form/validation/FormValidatorReCaptcha.inc.php [hunk 3]
```diff
@@ -56,59 +56,72 @@ function isValid() {
...
+						],
+						[
+							'name' => 'remoteip',
+							'contents' => $ip,
+						],
+					]
+				]
+			);
+		} catch (GuzzleHttp\Exception\RequestException $e) {
+			$response = $e->getResponse();
```
Comment (asmecher) [inline classes/form/validation/FormValidatorReCaptcha.inc.php:115]: Why not just leave the `RequestException` to bubble up to calling code? It seems odd to me to catch it here then throw an `Exception` just below (if I'm reading this right).

If this is done so that all thrown exceptions can use locale keys, maybe introduce a `LocalizedException` class (or a trait) to permit exceptions to be instantiated with locale keys that can be translated when the message is fetched? Just a thought.

File: classes/webservice/WebService.inc.php [hunk 4]
```diff
@@ -1,349 +0,0 @@
-<?php
```
Comment (asmecher) [inline classes/webservice/WebService.inc.php:1]: I'm happy to see these unused files go, but could you submit that as a separate pull request?

**Review by asmecher — CHANGES_REQUESTED** [review]
Just a few comments/changes, @jonasraoni -- generally this looks good!

We're in a feature freeze for 3.3 now, so this would be merged to `master` after `stable-3_3_0` is forked.

File: templates/frontend/pages/userLogin.tpl (hunk 1 above)
Comment (jonasraoni) [reply templates/frontend/pages/userLogin.tpl:82]: I'll replace another occurrence of it in the codebase.

File: classes/form/validation/FormValidatorReCaptcha.inc.php (hunk 3 above)
Comment (jonasraoni) [reply classes/form/validation/FormValidatorReCaptcha.inc.php:115]: Here I just wanted to have a custom friendly message going out. But in general I think it's better to let the unexpected/unrecoverable errors leak and treat in the calling site (to decrease the pollution). So I'll update :)

I think the LocalizedException sounds good, just to make it clear of what's inside it... I see it as a base exception that we should use whenever is possible.

The trait would be great to extend existing exceptions with an embedded logic, but the instanceof will not work with it. As far as I know, there's just the [class_uses](https://www.php.net/class_uses).

Also, in ancient times I used to have an `UserException`, just a lazy way to specify that the message is friendly enough to be displayed to the user without special treatments. Not sure if it's a good approach 😁

Comment (NateWr) [reply classes/form/validation/FormValidatorReCaptcha.inc.php:115]: One note on the `LocalizedException` thing: we need to move away from relying entirely on locale keys as error codes. The current authorization system has this problem, where you can set a locale key as an authorization message. But there's no support for params in localized messages (eg - "An error with submission {$id}."), so it can be challenging to return good error messages to users.

I think that, in general, when an error is thrown deep in the code stack, it should only be converted into a user-facing message by the handler. Only the handler knows who they are speaking to, so the same error might result in a different message depending on what the request is.

For example, the REST API may return a more technical message while in other cases we may want to use a more user-friendly message. Or, another example, when uploading a file there may be a number of different reasons for failure. We want to group a bunch of them, like permission or server configuration issues, in a generic message to the user ("Your file could not be uploaded because of a system error. Please contact administrative support.") while showing better error messages when the user needs to take an action ("Your file is too big").

Comment (jonasraoni) [reply classes/form/validation/FormValidatorReCaptcha.inc.php:115]: Hi @NateWr, this topic is debatable, some programming languages even have base `Exception` classes with support for localization. Also, we could add support for the parameters =]

I think exceptions are meant for developers. But customizing error messages might be tedious, as it will require some custom Exceptions and possibly duplicated treatments. So, follow 2 points:
1. Is it worth to localize error messages for developers? I personally prefer to read everything in English, since the code is already in English and it's easier to Google for solutions...
2. Should we give as much information as we can to the user? Or just customize errors when the user has an opportunity to fix by himself? I think the latter, which you mentioned, is better... But the former is useful for beginners using an application, trying to find what's going on 🤔

Regardless of the answers, I think we need to add loggers, to avoid hiding the exceptions under the carpet.

Comment (asmecher) [reply classes/form/validation/FormValidatorReCaptcha.inc.php:115]: Generally I think we should adopt PHP's SPL exceptions rather than doing a lot of subclassing -- but as to the matter of presenting friendly messages to users while collecting richer details in the log, I don't know what's currently best practice. I'm sure Laravel will have a decent pattern for this -- do either of you know what they do? (The "your file is too big" example is a good one.)

Comment (jonasraoni) [reply classes/form/validation/FormValidatorReCaptcha.inc.php:115]: Yep, I agree. About subclassing, I think we should avoid (mostly to avoid pollution) until it's really needed for a customization/logic. The extra information to help on debugging can be just logged...

In my case I don't need extra classes, I might display just a generic "system error, try again" or "validation failed"... But the "system error" should be logged somewhere with a granularity (invalid configuration = error, request failed = warning).

Comment (jonasraoni) [reply classes/form/validation/FormValidatorReCaptcha.inc.php]: Well, I've replaced the localized exceptions by texts in English and just displayed a generic message to the user stating that the captcha validation failed.

Comment (asmecher) [reply classes/form/validation/FormValidatorReCaptcha.inc.php]: Thanks, @jonasraoni, that's a good solution!

**Comment by jonasraoni** [comment]
@asmecher As the branch is created, I think it's ok to merge this one.

**Comment by asmecher** [comment]
Merged, thanks!
